### PR TITLE
Refactor how deployment works

### PR DESCRIPTION
**Pull request by Others**
This boy is a chonker. I pulled out all the deployment logic into
a `deployment.ActionManager` object. It's a lot cleaner and now
handles a lot of cases we weren't handling before. (Like startup!)

It isn't perfect and the error handling is probably problematic,
so let's call this WIP. I also fixed a lot of little nits that
were bothering me during the refactor so the diff is kinda
unnecessarily large.

Really looking for feedback on this one!

### CODE REVIEW

**Comment by Others** [comment]
@pcodes This is kinda a big change, can you also do a scan through for me?

File: activator/git.go [hunk 1]
```diff
@@ -12,8 +12,8 @@ import (
 )
 
 //Checkout head of specific repo
-func checkoutHead(path string) error {
-	cmd := exec.Command("git", "checkout", "HEAD")
+func checkout(path string, hash string) error {
+	cmd := exec.Command("git", "checkout", hash)
```
Comment (pcodes) [inline activator/git.go:16]: Why have we switched away from just cloning the HEAD?

File: database/database.go [hunk 2]
```diff
@@ -40,7 +41,7 @@ func (driver *Driver) FindUserID(githubUsername string) (string, error) {
 	return userID, nil
 }
 
-func (driver *Driver) FindComponentID(compID *worker.ComponentID) (string, error) {
+func (driver *Driver) FindComponentID(compID worker.ComponentID) (string, error) {
```
Comment (pcodes) [inline database/database.go:44]: Why are these no longer pointers?

File: database/database.go [hunk 3]
```diff
@@ -75,8 +76,53 @@ func (driver *Driver) FindWorkerID(workerName string) (string, error) {
...
+func (driver *Driver) SetWorkerRunningComponents(workerID string, compIDs []worker.ComponentID) error {
+	var dbCIDs = make([]string, 0)
+
+	for _, id := range compIDs {
+		dbCID, err := driver.FindComponentID(id)
+		if err != nil {
+			return err
+		}
+		dbCIDs = append(dbCIDs, dbCID)
+	}
```
Comment (pcodes) [inline database/database.go:88]: What is this doing?

File: database/database.go [hunk 4]
```diff
@@ -75,8 +76,53 @@ func (driver *Driver) FindWorkerID(workerName string) (string, error) {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 func (driver *Driver) InsertStats(workerID string, componentStatus worker.ComponentStats) error {
-	compID, err := driver.FindComponentID(&componentStatus.ID)
+	compID, err := driver.FindComponentID(componentStatus.ID)
```
Comment (pcodes) [inline database/database.go:125]: Why this change?

File: database/database.go [hunk 5]
```diff
@@ -183,3 +228,58 @@ func (driver *Driver) PurgeDeploymentEntry(compID *worker.ComponentID) error {
...
+	deleteQuery := `DELETE FROM v9.public.deploying`
+	_, err := driver.db.Exec(deleteQuery)
+	if err != nil {
+		return fmt.Errorf("could not delete from deploying table: %w", err)
+	}
+
+	return nil
+}
+
+func (driver *Driver) FindActiveComponents() ([]worker.ComponentPath, error) {
```
Comment (pcodes) [inline database/database.go:242]: Does the use of this function still work with the new method we have for "active" vs "pending" components? Or has that never mattered and only a front-end concept

File: deployment/actionManger.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+				return err
+			}
+		}
+	}
+
+	log.Info.Println("Finished dirty state handling")
+	return nil
+}
+
+func (mgr *ActionManager) deactivateNonactive(w *worker.V9Worker, active []worker.ComponentPath) error {
```
Comment (pcodes) [inline deployment/actionManger.go:174]: I know nonactive is a word but it looks weird

File: deployment/actionManger.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+			return err
+		}
+
+		// If it is running somewhere we are good
+		if status.ContainsPath(path) {
+			return nil
+		}
+	}
+
+	// Otherwise pick a worker randomly and deploy there
```
Comment (pcodes) [inline deployment/actionManger.go:211]: Add TODO for load balancing?

File: handlers/apiHandler.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	err = h.driver.SetDeploymentIntention(p.ID, p.NewDeploymentIntention)
+	if err != nil {
+		log.Error.Println("Failed to update status on database", err)
+		return
+	}
+	// Notify Action Manager
+	h.actionManager.NotifyComponentStateChanged()
+	// Send Response
+
+	_, err = fmt.Fprintf(w, "10/4 Good Buddy")
```
Comment (pcodes) [inline handlers/apiHandler.go:57]: lol

File: handlers/pushHandler.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	return &handler
+}
+
+func (h *PushHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// TODO: Ensure we only deploy from master
+
+	// Load secret from env
+	secret, exists := os.LookupEnv("GITHUB_SECRET")
+	if !exists {
+		log.Error.Println("Failed to find Github secret")
```
Comment (pcodes) [inline handlers/pushHandler.go:34]: I think we should push the secret to github so we can always find it

File: handlers/pushHandler.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		user = payload.Installation.Account.Login
+		for _, repo := range payload.RepositoriesAdded {
+			compID := worker.ComponentID{User: user, Repo: repo.Name, Hash: hash}
+			h.processComponentEvent(compID)
+		}
+	default:
+		parsedPayload := payload.(github.PushPayload)
+		user = parsedPayload.Repository.Owner.Login
+		repo = parsedPayload.Repository.Name
+		// NOTE: We don't know that the branch is correct, so we can't trust the head commit
```
Comment (pcodes) [inline handlers/pushHandler.go:73]: I'm seeing this now after asking my initial question, would love to chat about this

**Review by pcodes — CHANGES_REQUESTED** [review]
There's a subtle bug in the ActionManager. Can you find it?

File: activator/git.go (hunk 1 above)
Comment (Others) [reply activator/git.go:16]: We uses hashes throughout for version consistency. If you pause and unpause, it makes sure to deploy the same commit (unless some event from Github has come along). I think the action manager might get confused if the deployment system just sends random hashes -- but I'm not 100% sure.

File: database/database.go (hunk 2 above)
Comment (Others) [reply database/database.go:44]: Pointers let you modify the original object. I thought that was scary and we don't use that functionality. So I removed the pointers

File: database/database.go (hunk 3 above)
Comment (Others) [reply database/database.go:88]: It's just finding database component ids and storing them in `dbCIDs`

File: database/database.go (hunk 4 above)
Comment (Others) [reply database/database.go:125]: Goes with the other pointer removal change

File: database/database.go (hunk 5 above)
Comment (Others) [reply database/database.go:242]: pending is a front end concept

File: deployment/actionManger.go (hunk 6 above)
Comment (Others) [reply deployment/actionManger.go:174]: Yeah it was a little painful to write this haha -- is there a better name you can think of?

File: deployment/actionManger.go (hunk 7 above)
Comment (Others) [reply deployment/actionManger.go:211]: @hjaensch7 is there an issue?

File: database/database.go (hunk 3 above)
Comment (pcodes) [reply database/database.go:88]: What is a CID is more of my question

**Comment by Others** [comment]
The subtle bug is the following:
- Timmy deploys a component from his repo `timmysGreatComponent`
- Timmy pushes a commit to his `dev` branch
- The deployment manager sees it should deploy (since there is a new push event)
- There are now two workers running identical versions of the component
- This situation remains until there is a push to `master`
- (so we could end up with timmy's component on every worker)

I think this means we need to be cleverer about push events -- make sure they come from the right branch or stop using the `HEAD` trick. We also could just clean up this situation in the auto scaling stack, but that seems inelegant.

@pcodes @hjaensch7

File: database/pollingPopulator.go [hunk 11]
```diff
@@ -35,13 +35,23 @@ func (populator *PollingPopulator) pollWorkersToDatabase() {
 
 		// TODO: Clear out old stats
```
Comment (hjaensch7) [inline database/pollingPopulator.go:36]: Issue?

File: deployment/actionManger.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+	notRunningAnyVersion := make([]*worker.V9Worker, 0)
+
+	for _, w := range mgr.workers {
+		status, err := w.Status()
+		if err != nil {
+			return err
+		}
+
+		for _, runningComp := range status.ActiveComponents {
+			// If we find someone running exactly this ID, we have ensured some worker is running this OD
```
Comment (hjaensch7) [inline deployment/actionManger.go]: nit: "OD" should be "ID"

File: deployment/actionManger.go [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+	default:
+	}
+}
+
+func (mgr *ActionManager) UpdateComponentHash(compID worker.ComponentID) {
+	mgr.pathHashUpdater <- compID
+}
+
+func (mgr *ActionManager) HandleDirtyState() error {
+	// TODO: Parallelize this step (it basically single threads the deployment manager at the moment)
```
Comment (hjaensch7) [inline deployment/actionManger.go:89]: issue?

**Review by hjaensch7 — COMMENTED** [review]
This looks great. It's a big refactor so we should be sure that it's tested well. Just left a couple nits and marked some potential issues.

File: database/database.go (hunk 3 above)
Comment (Others) [reply database/database.go:88]: component id

File: database/pollingPopulator.go (hunk 11 above)
Comment (Others) [reply database/pollingPopulator.go:36]: #34

File: deployment/actionManger.go (hunk 13 above)
Comment (Others) [reply deployment/actionManger.go:89]: #33

**Comment by Others** [comment]
Pushed changes, also made issues for all the problems we need to resolve in future

**Comment by Others** [comment]
@pcodes @hjaensch7 ready for re-review, should be ready to merge

**Review by hjaensch7 — APPROVED** [review]
Looks good, make a BUG issue for the infinite deploy bug

**Review by pcodes — APPROVED** [review]
So good I didn't even need to read it. Keep up the good documentation!

**Pull request merged**